File: frontend/src/features/character-sheet/systems/vtm_v5/schema.ts
import { z } from "zod";

export const VTM_PHYSICAL_ATTRS = ["str", "dex", "sta"] as const;
export const VTM_SOCIAL_ATTRS = ["cha", "man", "com"] as const;
export const VTM_MENTAL_ATTRS = ["int", "wil", "cer"] as const;

export const VTM_ATTRIBUTES = [
  ...VTM_PHYSICAL_ATTRS,
  ...VTM_SOCIAL_ATTRS,
  ...VTM_MENTAL_ATTRS,
] as const;

export type VtmAttribute = (typeof VTM_ATTRIBUTES)[number];

export const VTM_ATTRIBUTE_LABELS: Record<VtmAttribute, string> = {
  str: "Fuerza",
  dex: "Destreza",
  sta: "Vigor",
  cha: "Carisma",
  man: "Manipulación",
  com: "Compostura",
  int: "Inteligencia",
  wil: "Voluntad",
  cer: "Astucia",
};

export const VTM_SKILL_CATEGORIES = ["physical", "social", "mental"] as const;
export type VtmSkillCategory = (typeof VTM_SKILL_CATEGORIES)[number];

export const VTM_SKILL_CATEGORY_LABELS: Record<VtmSkillCategory, string> = {
  physical: "Física",
  social: "Social",
  mental: "Mental",
};

const attributeScore = z.number().int().min(1).max(5);

const skillSchema = z.object({
  name: z.string().min(1, "Nombre requerido"),
  category: z.enum(VTM_SKILL_CATEGORIES),
  dots: z.number().int().min(0).max(5),
});

const disciplineSchema = z.object({
  name: z.string().min(1, "Nombre requerido"),
  level: z.number().int().min(1).max(5),
});

export const vtmV5SheetSchema = z.object({
  attributes: z.object({
    str: attributeScore,
    dex: attributeScore,
    sta: attributeScore,
    cha: attributeScore,
    man: attributeScore,
    com: attributeScore,
    int: attributeScore,
    wil: attributeScore,
    cer: attributeScore,
  }),
  skills: z.array(skillSchema),
  disciplines: z.array(disciplineSchema),
  health: z.object({
    superficial: z.number().int().min(0),
    aggravated: z.number().int().min(0),
    max: z.number().int().min(1),
  }),
  hunger: z.number().int().min(0).max(5),
});

export type VtmV5Sheet = z.infer<typeof vtmV5SheetSchema>;

export function defaultVtmV5Sheet(): VtmV5Sheet {
  return {
    attributes: {
      str: 1,
      dex: 1,
      sta: 1,
      cha: 1,
      man: 1,
      com: 1,
      int: 1,
      wil: 1,
      cer: 1,
    },
    skills: [],
    disciplines: [],
    health: { superficial: 0, aggravated: 0, max: 7 },
    hunger: 1,
  };
}

export function parseVtmV5Sheet(raw: unknown): VtmV5Sheet {
  const result = vtmV5SheetSchema.safeParse(raw);
  if (result.success) return result.data;
  return defaultVtmV5Sheet();
}

File: backend/app/rules/vtm_v5/schema.py
from pydantic import BaseModel, ConfigDict, Field

VTM_ATTRIBUTES = (
    "str",
    "dex",
    "sta",
    "cha",
    "man",
    "com",
    "int",
    "wil",
    "cer",
)

VTM_SKILL_CATEGORIES = ("physical", "social", "mental")

CATEGORY_DEFAULT_ATTRIBUTE: dict[str, str] = {
    "physical": "dex",
    "social": "man",
    "mental": "int",
}


class AttributeBlock(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    str: int = Field(default=1, ge=1, le=5)
    dex: int = Field(default=1, ge=1, le=5)
    sta: int = Field(default=1, ge=1, le=5)
    cha: int = Field(default=1, ge=1, le=5)
    man: int = Field(default=1, ge=1, le=5)
    com: int = Field(default=1, ge=1, le=5)
    intelligence: int = Field(default=1, ge=1, le=5, validation_alias="int", serialization_alias="int")
    wil: int = Field(default=1, ge=1, le=5)
    cer: int = Field(default=1, ge=1, le=5)

    def score(self, attribute: str) -> int:
        key = attribute.strip().lower()
        if key not in VTM_ATTRIBUTES:
            raise ValueError(f"Unknown attribute: {attribute}")
        if key == "int":
            return self.intelligence
        return getattr(self, key)


class SkillEntry(BaseModel):
    name: str
    category: str = "physical"
    dots: int = Field(default=0, ge=0, le=5)


class DisciplineEntry(BaseModel):
    name: str
    level: int = Field(default=1, ge=1, le=5)


class HealthBlock(BaseModel):
    superficial: int = Field(default=0, ge=0)
    aggravated: int = Field(default=0, ge=0)
    max: int = Field(default=7, ge=1)


class VtmV5Sheet(BaseModel):
    attributes: AttributeBlock = Field(default_factory=AttributeBlock)
    skills: list[SkillEntry] = Field(default_factory=list)
    disciplines: list[DisciplineEntry] = Field(default_factory=list)
    health: HealthBlock = Field(default_factory=HealthBlock)
    hunger: int = Field(default=1, ge=0, le=5)


SUPPORTED_ROLL_TYPES = [
    "skill_check",
    "attribute_check",
    "rouse_check",
    "discipline_check",
]
